azkaban-common/src/main/java/azkaban/scheduler/MissedSchedulesManager.java

```diff
@@ -0,0 +1,216 @@
+      this.missedSchedulesExecutor.shutdown();
+      if (!this.missedSchedulesExecutor.awaitTermination(30, TimeUnit.SECONDS)) {
+        this.missedSchedulesExecutor.shutdownNow();
+      }
+    }
+    return true;
+  }
+
+  /**
+   * A MissedSchedule Task is a unit that store the critical information to execute a missedSchedule operation.
+   * A Task is capable to
+   * 1. notice user
+   * 2. execute back execution if enabled feature flag (TODO)
+   * when a missed schedule happens.
+   * */
+  public static class MissedSchedulesOperationTask implements Runnable {
+    private static final Logger LOG = LoggerFactory.getLogger(MissedSchedulesOperationTask.class);
+    private final Emailer emailer;
+    // emailRecipients to notify users the schedules missed
+    private final List<String> emailRecipients;
+    // the email body
+    private final String emailMessage;
+    // back execute action
+    // if user disabled the config, the action could be null.
+    private ExecuteFlowAction executeFlowAction;
+
+    public MissedSchedulesOperationTask(List<Long> missedScheduleTimes,
+        Emailer emailer,
+        List<String> emailRecipients,
+        String projectName,
+        String flowName) {
+      this.emailer = emailer;
+      this.emailRecipients = emailRecipients;
+      final String missScheduleTimestampToDate = formatTimestamps(missedScheduleTimes);
+      MessageFormat messageFormat = new MessageFormat("This is autogenerated email to notify to flow owners that"
+          + " the flow {0} in project {1} has scheduled the executions on {2} but failed to trigger on time.");
+      this.emailMessage = messageFormat.format(new Object[]{flowName, projectName, missScheduleTimestampToDate});
+      this.executeFlowAction = null;
+    }
+
+    public MissedSchedulesOperationTask(List<Long> missedScheduleTimes,
+        Emailer emailer,
+        List<String> emailRecipients,
+        ExecuteFlowAction executeFlowAction) {
+      this.emailer = emailer;
+      this.emailRecipients = emailRecipients;
+      this.executeFlowAction = executeFlowAction;
+      final String missScheduleTimestampToDate = formatTimestamps(missedScheduleTimes);
+      MessageFormat messageFormat = new MessageFormat("This is autogenerated email to notify to flow owners that"
+          + " the flow {0} in project {1} has scheduled the executions on {2} but failed to trigger on time. "
+          + "Back execution will start soon as you enabled the config.");
+      this.emailMessage = messageFormat.format(
+          new Object[]{executeFlowAction.getFlowName(), executeFlowAction.getProjectName(), missScheduleTimestampToDate});
+    }
+
+    private String formatTimestamps(List<Long> missScheduleTimestamps) {
+      List<String> datesFromTimestamps = missScheduleTimestamps.stream()
+          .map(timestamps -> new Date(timestamps).toString())
+          .collect(Collectors.toList());
+      return String.join(",", datesFromTimestamps);
```

Comment:
What's the timezone used for the conversion?
Please follow the same time related conventions as for other email notifications.

Reply:
apply the flowtrigger email date format. It will be on thread's timezone.

Reply:
applied with formatDateTime instead. still the server local time instead of UTC.